test/chart/lib/chartlib.sh

```diff
@@ -0,0 +1,477 @@
+#!/usr/bin/env bash
+
+# Copyright 2018 The Helm Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+shopt -s nullglob
+
+
+readonly REMOTE="${REMOTE:-origin}"
+readonly TARGET_BRANCH="${TARGET_BRANCH:-master}"
+readonly TIMEOUT="${TIMEOUT:-300}"
+readonly LINT_CONF="${LINT_CONF:-/testing/etc/lintconf.yaml}"
+readonly CHART_YAML_SCHEMA="${CHART_YAML_SCHEMA:-/testing/etc/chart_schema.yaml}"
+readonly VALIDATE_MAINTAINERS="${VALIDATE_MAINTAINERS:-true}"
+
+# Special handling for arrays
+[[ -z "${CHART_DIRS[*]}" ]] && CHART_DIRS=(charts); readonly CHART_DIRS
+[[ -z "${EXCLUDED_CHARTS[*]}" ]] && EXCLUDED_CHARTS=(); readonly EXCLUDED_CHARTS
+[[ -z "${CHART_REPOS[*]}" ]] && CHART_REPOS=(); readonly CHART_REPOS
+
+echo
+echo '--------------------------------------------------------------------------------'
+echo ' Environment:'
+echo " REMOTE=$REMOTE"
+echo " TARGET_BRANCH=$TARGET_BRANCH"
+echo " CHART_DIRS=${CHART_DIRS[*]}"
+echo " EXCLUDED_CHARTS=${EXCLUDED_CHARTS[*]}"
+echo " CHART_REPOS=${CHART_REPOS[*]}"
+echo " TIMEOUT=$TIMEOUT"
+echo " LINT_CONF=$LINT_CONF"
+echo " CHART_YAML_SCHEMA=$CHART_YAML_SCHEMA"
+echo " VALIDATE_MAINTAINERS=$VALIDATE_MAINTAINERS"
+echo '--------------------------------------------------------------------------------'
+echo
+
+
+# Detects chart directories that have changes against the
+# target branch ("$REMOTE/$TARGET_BRANCH").
+chartlib::detect_changed_directories() {
+    local merge_base
+    merge_base="$(git merge-base "$REMOTE/$TARGET_BRANCH" HEAD)"
+
+    local changed_dirs=()
+    local dir
+
+    while read -r dir; do
+        local excluded=
+        for excluded_dir in "${EXCLUDED_CHARTS[@]}"; do
+            if [[ "$dir" == "$excluded_dir" ]]; then
+                excluded=true
+                break
+            fi
+        done
+        if [[ -z "$excluded" && -d "$dir" ]]; then
+            changed_dirs=("${changed_dirs[@]}" "$dir")
+        fi
+
+    ## @munnerz: because the cert-manager repository stores charts in the contrib/
+    ## subdirectory, we must modify the below line from $1/$2 to be $1/$2/$3.
+    ## In future, we should PR upstream so we no longer hardcode the depth of
+    ## directories required for this script.
+    done < <(git diff --find-renames --name-only "$merge_base" "${CHART_DIRS[@]}" | awk -F/ '{ print $1"/"$2"/"$3 }' | uniq)
```

Comment:
This single line change is the reason why we can't just use the upstream docker image on its own for running the test 😢 

I'd love to get this fixed upstream! Right now though, we cannot merge changes to our Helm chart, so I have opted to vendor the bash library for the time being.

Reply:
/cc @unguiculus